src/RtAudioInterface.cpp

```diff
@@ -312,40 +359,109 @@ int RtAudioInterface::wrapperRtAudioCallback(void* outputBuffer, void* inputBuff
                                              unsigned int nFrames, double streamTime,
                                              RtAudioStreamStatus status, void* userData)
 {
-    return static_cast<RtAudioInterface*>(userData)->RtAudioCallback(
-        outputBuffer, inputBuffer, nFrames, streamTime, status);
+    RtAudioInterface* interface = static_cast<RtAudioInterface*>(userData);
+    return interface->RtAudioCallback(outputBuffer, inputBuffer, nFrames, streamTime,
+                                      status);
 }
 
 //*******************************************************************************
 void RtAudioInterface::RtAudioErrorCallback(RtAudioError::Type type,
                                             const std::string& errorText)
 {
     if ((type != RtAudioError::WARNING) && (type != RtAudioError::DEBUG_WARNING)) {
-        std::cout << '\n' << errorText << '\n' << std::endl;
+        std::cout << errorText << '\n' << std::endl;
         throw std::runtime_error(errorText);
+    } else if (type == RtAudioError::WARNING) {
+        std::cout << errorText << '\n' << std::endl;
+    } else if (type == RtAudioError::DEBUG_WARNING) {
+        std::cout << errorText << '\n' << std::endl;
     }
 }
 
 //*******************************************************************************
-int RtAudioInterface::startProcess() const
+int RtAudioInterface::startProcess()
 {
     try {
-        mRtAudio->startStream();
+        if (mRtAudio != NULL) {
+            mRtAudio->startStream();
+        }
     } catch (RtAudioError& e) {
-        std::cout << '\n' << e.getMessage() << '\n' << std::endl;
+        std::cout << e.getMessage() << '\n' << std::endl;
         return (-1);
     }
     return (0);
 }
 
 //*******************************************************************************
-int RtAudioInterface::stopProcess() const
+int RtAudioInterface::stopProcess()
 {
     try {
-        mRtAudio->closeStream();
+        if (mRtAudio != NULL) {
+            mRtAudio->closeStream();
+            AudioInterface::setDevicesWarningMsg("");
+            AudioInterface::setDevicesErrorMsg("");
+        }
     } catch (RtAudioError& e) {
-        std::cout << '\n' << e.getMessage() << '\n' << std::endl;
+        std::cout << e.getMessage() << '\n' << std::endl;
         return (-1);
     }
     return 0;
 }
+
+//*******************************************************************************
+void RtAudioInterface::getDeviceList(QStringList* list, bool isInput)
+{
+    list->clear();
+    list->append(QStringLiteral("(default)"));
+
+    std::vector<RtAudio::Api> apis;
+    RtAudio::getCompiledApi(apis);
```

Comment:
I noticed there's a very short but noticeable delay on your videos when switching devices. Does it make much of a difference if we do this in `setup()` vs each time the device list is retrieved? I imagine the list of APIs available to rtaudio won't change much during a session. If I understand what this function is supposed to do, it seems like it might only matter if you install a new audio backend while jacktrip is already open?

But if this isn't contributing to the delay, then nvm - seems safe to leave it in.

Reply:
I think on Windows, switching devices just is a little slower and laggy, even without these changes. Versus on Mac I've never really noticed any lag. So I'm not sure how much there is we can do about it without a substantial code refactor.

Reply:
Could you close the list independently from the "starting" of the AudioInterface?

Reply:
Hmm that might be a good way to make it a smoother experience. I'll look into it on Monday.

Reply:
Took a look at this - it might be possible but I don't think there's a quick and easy way to do it.